src/FromData.purs

```diff
@@ -16,22 +16,42 @@ import Data.Ratio (Ratio, reduce)
 import Data.Traversable (for, traverse)
 import Data.Tuple (Tuple(Tuple))
 import Data.Tuple.Nested (type (/\), (/\))
-import Prim.TypeError (class Fail, Text)
+import Data.UInt (UInt)
 import Data.Unfoldable (class Unfoldable)
-import Types.ByteArray (ByteArray)
-import Types.PlutusData (PlutusData(Bytes, List, Map, Integer))
+import Helpers (bigIntToUInt)
+import Prim.TypeError (class Fail, Text)
+import Types.ByteArray (ByteArray, byteArrayToHex)
+import Types.PlutusData (PlutusData(Bytes, Constr, List, Map, Integer))
 
 class FromData (a :: Type) where
   fromData :: PlutusData -> Maybe a
 
+instance FromData Void where
+  fromData _ = Nothing
+
 instance FromData Unit where
   fromData (List []) = Just unit
   fromData _ = Nothing
 
 instance FromData Boolean where
-  fromData (Integer n)
-    | n == BigInt.fromInt 0 = Just false
-    | n == BigInt.fromInt 1 = Just true
+  fromData (Constr n [])
+    | n == zero = Just false
+    | n == one = Just true
+    | otherwise = Nothing
+  fromData _ = Nothing
+
+instance FromData a => FromData (Maybe a) where
+  fromData (Constr n [ pd ]) = case fromData pd of
+    Just Nothing | n == one -> Just Nothing
+    Just (Just x) | n == zero -> Just (Just x) -- Just is zero-indexed by Plutus
+    _ -> Nothing
+  fromData _ = Nothing
+
+instance (FromData a, FromData b) => FromData (Either a b) where
+  fromData (Constr n [ pd ]) = case fromData pd of
```

Comment:
Does it call itself?

Reply:
Not sure what you mean? The `pd` is already a `PlutusData` by construction, so we just need to check where it is zero- or one-indexed, and whether it's a `Left` or `Right` inside the `Just` (where the `Just` comes from the `fromData` definition)

Reply:
I believe this code is not going to work. I added a test and it failed.

Could you please take a look at https://github.com/Plutonomicon/cardano-browser-tx/tree/klntsky/either-fromdata-bug ?

Hopefully I'm not misunderstanding something.

Reply:
@klntsky really good spot. I just realised I was overnesting the Just/Left, we just need one layer of Just to confirm a "successful" fromData. Where `one` or `zero` just represents Right or Left. Check out my latest push, this should work with your test hopefully.

Reply:
Also note I can't write
```
case fromData pd of
    Just e | n == one -> Just (Left e)
    Just x | n == zero -> Just (Right x)
 ```
 because this concretises the type of `a` to be the same as `b`, hence
 ```

instance (FromData a, FromData b) => FromData (Either a b) where
  fromData (Constr n [ pd ])
    | n == zero = maybe Nothing (Just <<< Left) (fromData pd)
    | n == one = maybe Nothing (Just <<< Right) (fromData pd)
  fromData _ = Nothing
 ```
 I also fixed `Maybe`, there's only one type here, so we don't need to worry about any type confusion

Reply:
Here's `Maybe` for your convenience:
```
instance FromData a => FromData (Maybe a) where
  fromData (Constr n [ pd ]) = case fromData pd of
    Just _ | n == one -> Just Nothing
    Just x | n == zero -> Just (Just x) -- Just is zero-indexed by Plutus
    _ -> Nothing
  fromData _ = Nothing
  ```

Reply:
Just to be overly explicit, if your example, toData gives `Constr zero [ Integer one ]` and mapping the `Integer` in the array should give `Just one`, not `Just (Left one)` like I originally had.